### DISCUSSION TITLE
Is there support for PDM I2S microphones?

### DISCUSSION THREAD

**Discussion by rmick** [pschatzmann/arduino-audio-tools#203] [Q&A] [answered]
I am using the M5stack-Core 2 module, which includes almost everything you could desire, including an I2S mic and I2S speaker. The only problem is that this mic works in PDM I2S mode and I can't see anyway to configure this option in the library.

I believe that the Espressif framework supports this functionality, so is it possible to add it as an option in your amazing library?

**Comment by pschatzmann** [comment, marked as answer]
Oh, I didn't know that this exists. I just committed a change to my project. 
In the I2S Config you can set now signal_type=PDM.
Please let me know if this works...

**Reply by rmick** [reply]
Thanks, that is awesome. :-)

**Reply by rmick** [reply]
Not sure if it works, because there are a couple of compilation error in other modules.

Arduino/libraries/arduino-audio-tools/src/AudioHttp/HttpRequest.h: In member function 'virtual int audio_tools::HttpRequest::connect(const char*, uint16_t, int32_t)':
Arduino/libraries/arduino-audio-tools/src/AudioTools/AudioLogger.h:129:135: error: format '%s' expects argument of type 'char*', but argument 4 has type 'int' [-Werror=format=]
 #define LOG_OUT(level, ...) snprintf(AudioLogger::instance().prefix(__FILE__,__LINE__, level).str(),LOG_PRINTF_BUFFER_SIZE,__VA_ARGS__); AudioLogger::instance().println();
                                                                                                                                       ^
Arduino/libraries/arduino-audio-tools/src/AudioTools/AudioLogger.h:132:77: note: in expansion of macro 'LOG_OUT'
 #define LOGI(...) if (AudioLogger::instance().level()<=AudioLogger::Info) { LOG_OUT(AudioLogger::Info, __VA_ARGS__);}
                                                                             ^
Arduino/libraries/arduino-audio-tools/src/AudioHttp/HttpRequest.h:173:13: note: in expansion of macro 'LOGI'
             LOGI("connected %s timeout %d", is_connected, timeout);

**Comment by pschatzmann** [comment]
Strange, I can't reproduce your errors. 
- Is this in Arduino or PlatformIO ?
- What ESP32 Arduino Version are you using
- What Sketch are you compiling
- In Arduino Settings Compiler Warnings: do you have the setting Default ?

**Comment by rmick** [comment]
Answers below

**Comment by rmick** [comment]
All working perfectly.

Here are the config settings for anyone who wants to try the microphone in the M5Stack-Core2 unit.

   auto config                        = i2sStream.defaultConfig(RX_MODE);
   config.signal_type            = PDM;
   config.i2s_format             = I2S_STD_FORMAT; 
   config.sample_rate          = 22050;
   config.channels                =     2;
   config.bits_per_sample  =    16;
   config.pin_bck                 =    12;
   config.pin_data                =    34;
   config.pin_ws                   =     0;

**Comment by pschatzmann** [comment]
I updated the documentation and added a section for PDM microphones: https://github.com/pschatzmann/arduino-audio-tools/wiki/External-ADC
